src/main/java/com/google/sps/recommendation/ProcessNearestNeighbors.java

```diff
@@ -0,0 +1,119 @@
+import org.deeplearning4j.models.embeddings.loader.WordVectorSerializer;
+import org.deeplearning4j.models.word2vec.wordstore.VocabCache;
+
+import org.nd4j.linalg.api.buffer.DataType;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.stream.IntStream;
+import com.google.sps.tool.ArrayUtil;
+import org.mapdb.*;
+
+/* This function of this file is to ONLY pre-process the nearest neigbors of 300,000 words.
+ * It will create/update a binary .db file. Time results and optimal batch size
+ * are based on running this file on a Google Pixelbook. Time result may change for
+ * different hardwares.
+ *
+ * Processing Time:
+ * On a Google Pixelbook, a batch size of 500 words results in the fastest runtime for
+ * computing the nearest neighbors, while avoiding OOMs. We may see some speedup by
+ * migrating from N4dj matrix operations to a numerical computing library like
+ * numpy.
+ *
+ * Time example:
+ * Each batch of 500 words takes about 60,000 - 80,000 ms.
+ * Processing 100 batches can take from 1 hour 30 minutes - 2 hours.
+ *
+ * Some underlying reasons for this slow time performance:
+ * 1) N4dj's behind the hood operations. For example, to slice a large matrix by a range of indices,
+ *    N4dj has to make a copy which is really slow, especially with a big matrix.
+ * 2) Each computation of a batch results in sorting the
+ *    cosine similarity values in each row of size 300,000.
+ * 3) There is some time factor with serializing the result into the .db file.
+ *    Note: We use MapDB API to serialize a HashMap Object, which utilizes Random Access Files (RAF)
+ */
+public class ProcessNearestNeighbors {
+  public static void main(String[] args) throws IOException {
+    String path =
+        Paths.get("/home/ngothomas/GoogleNews-vectors-negative300-SLIM.bin.gz").toString();
+    WordVectors vec = WordVectorSerializer.readWord2VecModel(path);
+    InMemoryLookupTable lookupTable = (InMemoryLookupTable) vec.lookupTable();
+    VocabCache lookupCache = (VocabCache) vec.vocab();
+
+    /* Create an array that maps index to word,
+     * where the index represents where the word is
+     * in the word vector matrix 300,000 x 300 */
+    String[] indexToWord = new String[300000];
+    for (Object i : lookupCache.words()) {
+      String word = (String) i;
+      int index = vec.indexOf(word);
+      indexToWord[index] = word;
+    }
+
+    /* Get raw word vectors 300,000 x 300 matrix (syn0) and normalize it (vecA)
+     * syn0 naming comes from the original Word2Vec paper */
+    INDArray syn0 = lookupTable.getSyn0();
+    INDArray vecA = syn0.divColumnVector(syn0.norm2(1));
+
+    /* There are about 300,000 words, so we want to iterate 600 times
+     * each iteration computes a batch of 500 words */
+    int rangeStart = 0;
+    int rangeEnd = 499;
```

Comment:
Use `rangeEnd = 500` and `IntStream.range` instead of `rangeClosed`. Then you don't need to add one to `rangeStart` on line 83 either.

Reply:
thanks